crypto/crypto_aesctr_aesni.c

```diff
@@ -51,7 +51,12 @@ crypto_aesctr_aesni_stream_wholeblocks(struct crypto_aesctr * stream,
 	/* How many blocks should we process? */
 	num_blocks = (*buflen) / 16;
 
-	for (i = 0; i < num_blocks; i++) {
+	/*
+	 * This is 'for (i = num_blocks; i > 0; i--)', but ensuring that the
```

Comment:
Is counting down better than counting up?

If we count up, we could use `*inbuf + i * 16` and similarly for `outbuf` in order to avoid updating them on every loop.

Reply:
Yes, counting down is better.  I added an extra variable `j = 0; ... j++` and saw no speed change (possibly because it was compiled out).  But testing the loop comparison with:
```
#if 0
        } while (i > 0);
#else
        } while (j < num_blocks);
#endif
```

showed 1695 for the `(i > 0)` [as expected], but 1489 for `(j < num_blocks)`.  That actually drops to 1467 when I try the `(*arr + j*16)` trick.  I would have expected the compiler to translate both loops to "counting down", since "jump not equal" is faster than "compare two variables, then maybe jump".

I wouldn't have expected `*inbuf += 16; outbuf += 16;` to have a performance effect, since those can be calculated while the SSE part is encrypting the cipherblock.

I also played around with
```
__m128i * out = *outbuf;
...
     _mm_storeu_si128(&out[j], bufsse);
```

and again saw values in the upper 1400s.

Reply:
Weird.  Oh well, the code is a bit ugly but I'll put up with it for the performance boost.